### ISSUE TITLE
Python 3 builds of NVDA: in a particular PDF pages are numbered from 0 in Adobe Reader DC

### ISSUE THREAD

**Issue by lukaszgo1** [nvaccess/nvda#10011]
### Steps to reproduce:
1. Open attached document in Adobe Reader dc.
2. Notice that current page is announced as 0 of 14.
### Actual behavior:
Current page number is announced as 0 which makes no sense.
### Expected behavior:
First page should be announced as page 1
#### NVDA installed/portable/running from source:
Running from source
#### NVDA version:
Latest Python 3 master as of 1th August 2019
#### Windows version:
Windows 7 x64
#### Name and version of other software in use when reproducing the issue:
Acrobat Reader DC Continuous Release | Version 2019.012.20035
#### Other information about your system:
With latest master of NVDA using Python 2 first page is properly announced as page 1, so it is a regression.
#### Does the issue still occur after restarting your PC?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
As stated above latest Master using Python 2 as an installed one, and there it works properly.
[the-loewner-equation-and-the-derivative-of-its-solution.pdf](https://github.com/nvaccess/nvda/files/3456792/the-loewner-equation-and-the-derivative-of-its-solution.pdf)

**Comment by lukaszgo1** [comment]
@leonardder Could you add this to the Update to Python 3 project as this is a regression?

**Comment by leonardder** [comment]
Thanks for poking me.

To make sure, on Python 2 builds of NVDA, is page 1 announced as 1 of 14 or 1 of 15? IN python 3 builds, is page 2 announced as page 1, page 3 as page 2, etc?

**Comment by lukaszgo1** [comment]
I am sorry, I was in a rush when submitting this issue and haven't tested thoroughly. Here is what happens:
- With Python 2 version of NVDA when this document gains focus,, or when you switch pages with Ctrl+page down the following is announced:
Page 13/14 DocumentPath\DocumentName, page 13 graphic page 12 document content.
It is important to note that first and last page announcements are coming from NVDA and the one in the middle from Acrobat (I can differentiate them, because Acrobat is in Polish and NVDA in English).
So the second page announcement made by NVDA is incorrect whereas the first is accurate and agrees with page number given by Acrobat.
- With Python 3 build of NVDA when switching pages only second incorrect announcement of page number is given. It might seems to be an improvement, because NVDA talks less, but it isn't for a few reasons:
1. As stated above the page number given in this way is incorrect.
2. When page to which you switched is empty there is no announcement at all.
3. Page number is announced after reading 'graphic' at the top of the page. (It is specific to this document, however when switching page I would expect page announcement to be made first).

So to summarize we have two separate issues here:
1. The second page number announcement from NVDA is and was incorrect.
2. The output when switching pages or gaining focus on Acrobat document changed when moving to Python 3, and in some situations it isn't an improvement.

**Comment by leonardder** [comment]
I wonder how python 3 could have touched this.
Anyway, I will look into this at some point before 2019.3 will be released.

**Comment by leonardder** [comment]
I found the cause of this.

NVDA can use two methods to get the page number/label. It gets the page number first, than tries to get a label for the page. If NVDA can't fetch a label, it will simply use the page number. As page numbers are zero based, 1 is added to the number.
However, the following is documented about fetching the label:

> 	/** Returns the page textual label associated with the given pageNum.
> 	If no textual label exists for pageNum, returns the string representing the 
> 	integer value pageNum. */

So, these last lines describe exactly what's happening here. The string representing the integer value of the page number is zero based.
In C++ it is pretty quirky to perform arithmetic operations on a string. Therefore, I propose doing the following:

1. In the vbuf backend, don't perform any page number processing, i.e. return the 0 based page number
2. Process the page number in normalizeControlField/normalizeFormatField in AdobeAcrobat_TextInfo, whatever applies, I'm not sure. If it is a string, just announce the string. If it is a number, ad 1 to it.

**Comment by leonardder** [comment]
@lukaszgo1: Do you think you're able to come up with a solution for this, based on my comments? In case you're not familiar with NVDAHelper, you will have to remove the + 1 at line 325 of adobeAcrobat.cpp.

To process the page number, add something along those lines to virtualBuffers.adobeAcrobat.AdobeAcrobatTextinfo

```
	def _normalizeControlField(self, attrs):
		if "page-number" in attrs:
			try:
				attrs["page-number"] = int(attrs["page-number"]) + 1
			except ValueError:
				pass   # A non integer string label
```

**Comment by leonardder** [comment]
@lukaszgo1: Do you think you're able to come up with a solution for this, based on my comments? In case you're not familiar with NVDAHelper, you will have to remove the + 1 at line 325 of adobeAcrobat.cpp.

To process the page number, add something along those lines to virtualBuffers.adobeAcrobat.AdobeAcrobatTextinfo

```
	def _normalizeControlField(self, attrs):
		if "page-number" in attrs:
			try:
				attrs["page-number"] = int(attrs["page-number"]) + 1
			except ValueError:
				pass   # A non integer string label
``Actually, thinking some more about this.

It might be better to rename the page-number attribute to _page-number in the cpp code while at it. This way, you can make sure that the unprocessed page number never leaks into the output.

**Comment by lukaszgo1** [comment]
I'll look into this. However even if it would be fixed the following would still be an issue:
> When page to which you switched is empty there is no announcement at all.
Page number is announced after reading 'graphic' at the top of the page. (It is specific to this document, however when switching page I would expect page announcement to be made first).

**Comment by lukaszgo1** [comment]
And not announcing page number for empty pages is a Python 3 regression by the way.

**Comment by Adriani90** [comment]
@lukaszgo1 I can reproduce this in NVDA 2019.3 Beta 4 but only when loading the entire document at start. Not reproducible when each page is loaded.
